lib/bolt/plugin/terraform.rb

```diff
@@ -65,15 +56,57 @@ def load_statefile(opts)
         raise Bolt::FileError.new("Could not load Terraform state file #{filename}: #{e}", filename)
       end
 
+      # Format the list of resources into a list of [name, attribute map]
+      # pairs. This method handles both version 4 and earlier statefiles, doing
+      # the appropriate munging based on the shape of the data.
+      def extract_resources(state)
+        if state['version'] >= 4
+          state.fetch('resources', []).flat_map do |resource_set|
+            prefix = "#{resource_set['type']}.#{resource_set['name']}"
+            resource_set['instances'].map do |resource|
+              instance_name = prefix
+              instance_name += ".#{resource['index_key']}" if resource['index_key']
+
+              [instance_name, resource['attributes']]
+            end
+          end
+        else
+          state.fetch('modules', {}).flat_map do |mod|
+            mod.fetch('resources', {}).map do |name, resource|
+              [name, resource.dig('primary', 'attributes')]
+            end
+          end
+        end
+      end
+
+      # Look up a nested value from the resource attributes. The key is of the
+      # form `foo.bar.0.baz`. For terraform statefile version 3, this will
```

Comment:
This seems reasonable. I'm not sure how many array elements might end up nested under keys (hopefully not so many that it is hard to calculate index by "eye"). It was kind of nice in version 3 that the index were part of the path. We should add a note about this in the docs as my impression is that this is the most noticeable change for how we are resolving values.